frontend/packages/dev-console/src/components/pipelineruns/logs/logs-utils.ts

```diff
@@ -0,0 +1,142 @@
+import { TaskRuns } from '../../../utils/pipeline-augment';
+
+const getSortedContainerStatus = (
+  containers: ContainerSpec[],
+  containerStatuses: ContainerStatus[],
+): ContainerStatus[] => {
+  const containerNames = containers.map((c) => c.name);
+  const sortedContainerStatus = [];
+  containerStatuses.forEach((cs) => {
+    const containerIndex = containerNames.indexOf(cs.name);
+    sortedContainerStatus[containerIndex] = cs;
+  });
+  return sortedContainerStatus;
+};
+
+export const getRenderContainers = (
+  pod: PodKind,
+): { containers: ContainerSpec[]; stillFetching: boolean } => {
+  const containers: ContainerSpec[] = pod.spec?.containers ?? [];
+  const containerStatuses: ContainerStatus[] = pod.status?.containerStatuses ?? [];
+
+  const sortedContainerStatuses = getSortedContainerStatus(containers, containerStatuses);
+
+  const firstRunningCont = sortedContainerStatuses.findIndex(
+    (container) => containerToLogSourceStatus(container) !== LOG_SOURCE_TERMINATED,
+  );
+  return {
+    containers: containers.slice(
+      0,
+      firstRunningCont === -1 ? containers.length : firstRunningCont + 1,
+    ),
+    stillFetching: firstRunningCont !== -1,
+  };
+};
+
+const getOrderedStepsFromPod = (name: string, ns: string): Promise<ContainerStatus[]> => {
+  return k8sGet(PodModel, name, ns)
+    .then((pod: PodKind) => {
+      return getSortedContainerStatus(
+        pod.spec.containers ?? [],
+        pod.status?.containerStatuses ?? [],
+      );
+    })
+    .catch((err) => {
+      errorModal({ error: err.message || 'Error downloading logs.' });
+      return [];
+    });
+};
+
+type StepsWatchUrl = {
+  [key: string]: {
+    name: string;
+    steps: { [step: string]: string };
+  };
+};
+
+export const getDownloadAllLogsCallback = (
+  sortedTaskRuns: string[],
+  taskRunFromYaml: TaskRuns,
+  namespace: string,
+  pipelineRunName: string,
+): (() => void) => {
+  const getWatchUrls = (): Promise<StepsWatchUrl> =>
+    sortedTaskRuns.reduce(async (acc, currTask) => {
+      const { pipelineTaskName, status = {} } = taskRunFromYaml[currTask];
+      const { podName } = status;
+      const steps = await getOrderedStepsFromPod(podName, namespace);
+      const allStepUrls = steps.reduce((stepUrls, currentStep) => {
+        const { name } = currentStep;
+        if (containerToLogSourceStatus(currentStep) !== LOG_SOURCE_TERMINATED) return stepUrls;
+        const urlOpts = {
+          ns: namespace,
+          name: podName,
+          path: 'log',
+          queryParams: {
+            container: name,
+            follow: 'true',
+          },
+        };
+        return { ...stepUrls, [name]: resourceURL(PodModel, urlOpts) };
+      }, {});
+      const accumulator = await acc;
+      accumulator[currTask] = {
+        name: pipelineTaskName,
+        steps: { ...allStepUrls },
+      };
+      return Promise.resolve(accumulator);
+    }, Promise.resolve({}));
```

Comment:
```suggestion
    const stepsList: ContainerStatus[][] = await Promise.all(
      sortedTaskRuns.map((currTask) => {
        const { status = {} } = taskRunFromYaml[currTask];
        const { podName } = status;
        return getOrderedStepsFromPod(podName, namespace);
      }),
    );

    return sortedTaskRuns.reduce(async (acc, currTask, i) => {
      const { pipelineTaskName, status = {} } = taskRunFromYaml[currTask];
      const { podName } = status;
      const steps = stepsList[i];

      const allStepUrls = steps.reduce((stepUrls, currentStep) => {
        const { name } = currentStep;
        const currentStatus = containerToLogSourceStatus(currentStep);
        if (currentStatus === LOG_SOURCE_WAITING) return stepUrls;

        const urlOpts = {
          ns: namespace,
          name: podName,
          path: 'log',
          queryParams: {
            container: name,
            follow: 'true',
          },
        };
        return {
          ...stepUrls,
          [name]: {
            status: currentStatus,
            url: resourceURL(PodModel, urlOpts),
          } as WatchURLStatus,
        };
      }, {});

      acc[currTask] = {
        name: pipelineTaskName,
        steps: { ...allStepUrls },
      };
      return acc;
    }, {});
```

I am unsure what you're doing in this code, so I rewrote it when I was in the area to see if I could make sense of it. You are waiting (`await`) on the PREVIOUS accumulator? Not sure why - might have been to satisfy TypeScript types and the following issues you ran into.

Personally, I am not a huge fan of `async/await` array functions as I find it breaks the flow of the readable code. I'd prefer "get me all my network data" then "create my data", rather than doing them together. We're talking milliseconds loss in performance (if even that)...

You don't have to go with what I have here, but please fix it up so you're not waiting on the accumulator from the previous iteration.